Handle literal ranges the same way we handle other literal generics

Morriar opened the pull request:
### Motivation

Currently, literal ranges are desugared into calls to `Range.new` for historical reasons:

> @elliottt:
> yes, i think it's sort of accidental that Range.new has a more precise type right now
> looking at when it was first introduced, it was before generics existed in sorbet
> so i think it was just code-modded forwards, while types like Array and Hash got more attention

This means that the type of literal ranges is determined by the `Range.new` [signature](https://github.com/sorbet/sorbet/blob/master/rbi/core/range.rbi#L108-L115) which uses the type of the first argument.

This behaviour is 1) incoherent with other bare generics initialized through `new`: 

```ruby
T.reveal_type(Hash.new(a: 1, b: 2)) # Revealed type: T::Hash[T.untyped, T.untyped]
T.reveal_type(Array.new(1, 2)) # Revealed type: T::Array[T.untyped]
T.reveal_type(Range.new(1, 2)) # Revealed type: T::Range[Integer]
```

And 2) it's also brittle and leads to wrong types being inferred:

```ruby
T.reveal_type(Range.new(1)) # Revealed type: T::Range[Integer(1)]
```

^ it's not a range of literal `1`s, it's a `T::Range[Integer]`.
 
This pull request changes the way we handle literal ranges to use a similar approach than for literal arrays and hashes:

1. literal ranges are kept in the tree instead of being desugared away as `Range.new`
2. The range initialization is rewritten into a `Magic#buildRange`
3. The magic is used to attach an intrinsic for the return type figuring the return type

There is also two changes related to ranges:
1. Removing the duplicated signature for `Range::new` and keeping only `Range#initialize`
2. Fixing the test for incorrect `T::Range[Integer(1)]`

This will also be fixing the behaviour of the `T::Range[NilClass]` discussed in https://github.com/sorbet/sorbet/pull/3313#discussion_r459127000 once https://github.com/sorbet/sorbet/pull/3313 is merged.

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

See included automated tests.

elliottt commented on ast/Trees.h:
```diff
@@ -783,6 +784,25 @@ TREE(Array) : public Expression {
+    virtual void _sanityCheck();
+};
+CheckSize(Range, 56, 8);
+
```
I don't think you need the Range tree at all: you've defined the `buildRange` intrinsic on the `<Magic>` symbol already, and should just translate range syntax directly into that intrinsic in Desugar.cc.

Adding a new tree node is a really heavyweight solution, and will touch most passes. It would be great to avoid this if possible.

elliottt commented on ast/desugar/Desugar.cc:
```diff
@@ -1278,20 +1278,16 @@ TreePtr node2TreeImpl(DesugarContext dctx, unique_ptr<parser::Node> what) {
-                auto send = MK::Send2(loc, std::move(range), core::Names::new_(), std::move(from), std::move(to));
-                result = std::move(send);
+                auto range = MK::Range(loc, std::move(from), std::move(to), false);
+                result = std::move(range);
```
Here's where it should construct the call to the buildRange intrinsic, instead of constructing the Range node. We do something similar with literals for arrays, hashes, and string interpolation.

https://github.com/sorbet/sorbet/blob/33a668e7238aaab1f4164cba4c358b8dd529b5b3/ast/desugar/Desugar.cc#L144-L145

Morriar commented:
@elliottt from our discussion I though we wanted them to be handled exactly the same way than other generics.

> Could you remove the Range tree node, and switch to constructing the buildRange intrinsic directly in Desugar.cc

Yes, I dropped that commit and updated the expected desugar trees accordingly. The rest is unchanged.

elliottt commented:
> @elliottt from our discussion I though we wanted them to be handled exactly the same way than other generics.
What I had meant there was that we should rely on the use of the constructor from `T::Range[...]` to propagate type information. I don't think that I made the approach I was thinking of clear enough, sorry for leading you down the wrong path there!

> Yes, I dropped that commit and updated the expected desugar trees accordingly. The rest is unchanged.
Thanks! I'll re-review this today.

elliottt commented on core/types/calls.cc:
```diff
@@ -1363,6 +1363,26 @@ class Magic_buildArray : public IntrinsicMethod {
+class Magic_buildRange : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        if (args.args.empty()) {
```
Should this be `args.args.size() != 2`?

Morriar replied:
The size can be `1` for `(x..)`. `0` should never happen as `(..)` is a parse error.

This was confusing, I changed it for `ENFORCE(!args.args.empty(), "Magic_buildRange called without argument");`

elliottt commented on core/types/calls.cc:
```diff
@@ -1363,6 +1363,26 @@ class Magic_buildArray : public IntrinsicMethod {
+
+        auto type_member = Types::dropLiteral(args.args[0]->type);
+
+        if (type_member->isNilClass()) {
+            type_member = Types::untypedUntracked();
```
I'm not sure this case is necessary -- why not just use the if below that checks if it's a subtype of `NilClass` instead?

elliottt replied:
Or, why not always dropNil if it's a subtype of `NilClass`, and then return untypedUntracked if it collapses to bottom?

Morriar replied:
Good idea 👍

elliottt commented on core/types/calls.cc:
```diff
@@ -1363,6 +1363,26 @@ class Magic_buildArray : public IntrinsicMethod {
+            type_member = Types::dropNil(gs, type_member);
+        }
+
+        res.returnType = Types::rangeOf(gs, type_member);
```
After this lands, will your other PR update to check the type of the second arg in the event that the first is nil?

Morriar replied:
Yes, as this won't even parse at this stage.

I added an intermediary version to check the type of the second argument and return `T::Range[T.untyped]` if it's different from the first one.

elliottt commented on rbi/core/range.rbi:
```diff
@@ -102,18 +102,6 @@ class Range < Object
   extend T::Generic
   Elem = type_member(:out)
 
-  # Constructs a range using the given `begin` and `end`. If the `exclude_end`
-  # parameter is omitted or is `false`, the range will include the end object;
-  # otherwise, it will be excluded.
-  sig do
-    type_parameters(:U).params(
-      from: T.type_parameter(:U),
-      to: T.nilable(T.type_parameter(:U)),
-      exclude_end: T::Boolean
-    ).returns(T::Range[T.type_parameter(:U)])
-  end
-  def self.new(from, to, exclude_end=false); end
-
```
👍

elliottt commented on core/types/calls.cc:
```diff
@@ -1363,6 +1363,34 @@ class Magic_buildArray : public IntrinsicMethod {
+class Magic_buildRange : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        ENFORCE(!args.args.empty(), "Magic_buildRange called without argument");
```
I think this is reasonable, given that there's no user-visible way to call this with zero args.

elliottt commented on core/types/types.cc:
```diff
@@ -309,6 +320,11 @@ bool Type::isUntyped() const {
     return t != nullptr && t->symbol == Symbols::untyped();
 }
 
+bool Type::isNilClass() const {
+    auto *t = cast_type<ClassType>(this);
+    return t != nullptr && t->symbol == Symbols::NilClass();
+}
+
```
Is this needed anymore?

Morriar replied:
No it's not.

elliottt commented on test/testdata/desugar/range.rb:
```diff
@@ -8,15 +8,18 @@ def foo
-    T.reveal_type(d) # error: Revealed type: `T::Range[Integer]`
+    T.reveal_type(d) # error: Revealed type: `T::Range[T.untyped]`
     T.reveal_type(e) # error: Revealed type: `Integer`
     T.reveal_type(f) # error: Revealed type: `String`
```
Could you add a few cases that use `T::Range[...].new` as well?

Morriar replied:
I completely rewrote the test file to be more exhaustive on the three forms: `x..y`, `Range.new` and `T::Range[X].new`. Let me know if you want to add more test cases.

Morriar commented on core/types/calls.cc:
```diff
@@ -1363,6 +1363,34 @@ class Magic_buildArray : public IntrinsicMethod {
+
+        if (args.args.size() > 1) {
+            auto other = dropType(gs, args.args[1]->type);
+            if (other != Types::untypedUntracked() && other != type_member) {
```
Actually, I think we need to go back to the version I was checking for `isNilClass(typeMember)`.

We can only reason about `NilClass`:
* if the `end` bound is `NilClass`: `(42..)`, we have an endless-range
* As discussed before a `(NilClass..NilClass)` should be `T.untyped`
* And in the next PR, if `begin` is `NilClass`: `(nil..42)` we have a begin-less range

Any other case, if we get `begin` or `end` == `T.untyped`, the return type should be `T::Range(T.untyped)` as we can't infer anything.

elliottt commented on core/types/calls.cc:
```diff
@@ -1363,6 +1363,31 @@ class Magic_buildArray : public IntrinsicMethod {
 
+class Magic_buildRange : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
+        ENFORCE(!args.args.empty(), "Magic_buildRange called without argument");
+
+        auto rangeElemType = Types::dropLiteral(args.args[0]->type);
+        if (!rangeElemType->isNilClass()) {
+            rangeElemType = Types::dropNil(gs, rangeElemType);
+        }
+
+        if (args.args.size() > 1) {
+            auto other = Types::dropLiteral(args.args[1]->type);
+            if (rangeElemType->isNilClass() && other->isNilClass()) {
+                rangeElemType = Types::untypedUntracked();
+            } else if (rangeElemType->isNilClass()) {
+                rangeElemType = Types::dropNil(gs, other);
+            } else if (other->isUntyped() || (!other->isNilClass() && !Types::equiv(gs, other, rangeElemType))) {
+                rangeElemType = Types::untypedUntracked();
+            }
+        }
+
+        res.returnType = Types::rangeOf(gs, rangeElemType);
+    }
```
```suggestion
    void apply(const GlobalState &gs, DispatchArgs args, const Type *thisType, DispatchResult &res) const override {
        ENFORCE(args.args.size() == 2, "Magic_buildRange called without argument");

        auto rangeElemType = Types::dropLiteral(args.args[0]->type);
        auto firstArgIsNil = Types::isSubType(gs, rangeElemType, Types::nilClass());
        if (!firstArgIsNil) {
            rangeElemType = Types::dropNil(gs, rangeElemType);
        }

        auto other = Types::dropLiteral(args.args[1]->type);
        auto secondArgIsNil = Types::isSubType(gs, other, Types::nilClass());
        if (firstArgIsNil) {
            if (secondArgIsNil) {
                rangeElemType = Types::untypedUntracked();
            } else {
                rangeElemType = Types::dropNil(gs, other);
            }
        } else if (other->isUntyped() || (!secondIsNil && !Types::equiv(gs, other, rangeElemType))) {
            rangeElemType = Types::untypedUntracked();
        }

        res.returnType = Types::rangeOf(gs, rangeElemType);
    }
```

* You can enforce that the buildRange intrinsic always has exactly two arguments because the only calls to it are constructed directly in `Desugar.cc` with two arguments
* Knowing that it has two arguments means that the conditional on the argument count at line 1376 will always be true
* You can check to see if the type is a subtype of `NilClass` instead of implementing a custom `isNilClass` function
* What do you think about raising an error in the final else-if case? I think that this would happen when the two bounds of the range are not `nil` or `T.untyped`, which would indicate that the bounds are ill-formed.

Morriar commented:
@elliottt I integrated your suggestions 👍 

About the error: it seems that real world examples with different bounds can be used correctly:

```ruby
(Time.now.utc - 2.hours)...Time.now.utc
                           ^^^^^^^^^^^^ # error: Range with incoherent bounds: Float and Time

 (Time.now .. Date::Infinity.new)
              ^^^^^^^^^^^^^^^^^^ # error: Range with incoherent bounds: Time and Date::Infinity
```

trevor-stripe commented:
> @elliottt I integrated your suggestions 👍
> 
> About the error: it seems that real world examples with different bounds can be used correctly:
> 
> ```ruby
> (Time.now.utc - 2.hours)...Time.now.utc
>                            ^^^^^^^^^^^^ # error: Range with incoherent bounds: Float and Time
> 
>  (Time.now .. Date::Infinity.new)
>               ^^^^^^^^^^^^^^^^^^ # error: Range with incoherent bounds: Time and Date::Infinity
> ```

In this case is there a subtyping relationship between the bounds? Maybe we can reformulate the use of `Types::equiv` to a more permissive subtyping relationship?

Morriar commented:
> In this case is there a subtyping relationship between the bounds? Maybe we can reformulate the use of Types::equiv to a more permissive subtyping relationship?

Let's discuss about this in another PR. I think the same test should be applied at least for `T::Range.new`, maybe also for `Range.new`.

elliottt commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_HuHJEeifJ8pnZ9

elliottt commented:
This will require a few internal changes to land -- I'll make them and then merge this PR.

elliottt commented on core/types/calls.cc:
```diff
@@ -1363,6 +1363,31 @@ class Magic_buildArray : public IntrinsicMethod {
+            } else {
+                rangeElemType = Types::dropNil(gs, other);
+            }
+        } else if (other->isUntyped() || (!secondArgIsNil && !Types::equiv(gs, other, rangeElemType))) {
```
```suggestion
        } else if (!secondArgIsNil) {
```

I think you can simplify the conditional here quite a bit.

elliottt commented:
Could you rebase this PR on master?

elliottt commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_HzB7Aga9620ibn

elliottt commented:
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build result here:

→ https://go/builds/bui_HzDovrFYTmKRrs

elliottt commented:
Passing build on stripe CI: https://go/builds/bui_HzEICOkVebYcTX
